design/API_requirements_and_design.md

```diff
@@ -0,0 +1,59 @@
+# Thin-Edge API requirements and design
+
+One should be able to build a thin-edge executable from tedge plugins crates that have been implemented independently:
+
+* All these tedge plugins crates depends on the tedge_api crate that defines how
+  to create, connect and run plugin instances.
+* Each tedge plugin crate can define its own types of messages. For instance, a
+  tedge_telemetry plugin can defines measurements, events, alarms and setpoints
+  messages.
+* To exchange typed messages, plugin crates need dependency relationship. For
+  instance, the tedge_json_over_mqtt plugin crate will have to depend on the
+  tedge_telemetry plugin crate and the tedge_mqtt crate.
+* The tedge_api crate defines only messages related to the runtime as Shutdown.
+* The implementation of the mechanisms defined by the tedge_api are provided by
+  a tedge_runtime crate.
+* The final executable is build as an assemblage of all these plugin crates and
+  a `main.rs` that defines which plugins have to be instantiated and then
+  launched.
```

Comment:
```suggestion
  a `main.rs` that defines which plugins can potentially be instantiated and
  launched.
```

Reply:
This is something we have indeed to discuss. Somehow I see this change (can be vs have to) as a great source of complexity.

* This is why I wrote "have to" and work in that direction with [my proposal](https://github.com/thin-edge/thin-edge.io/pull/1153).
* But I also appreciate the benefits of the more complex but flexible [approach you proposed with Marcel](https://github.com/thin-edge/thin-edge.io/pull/979)

I open to discuss this point further.

Reply:
Complexity for whom? As I believe that to be the most critical aspect of having such an abstraction.

To help the discussion here's some definitions of different people that might use TE:

- Framework Developers (us)
- Framework Users (people that develop plugins/custom deployment of TE)
- TE Users (people that use a built binary and have to configure/deploy it)
- End Users (people that interface with a product/deployment of TE)

Here is what each person gains/loses when supporting optional instantiation of components from my point of view:

- End Users do not see any direct change that is quantifiable
- TE Users get more flexibility on their deployments as a single compiled binary can be re-purposed. It would be seen as a swiss-army knife of edge configuration. Losses are slightly larger binary sizes; more complexity to understand what component does what and how to configure them.
- Framework Users do not see any direct change as writing a component will still be done through the interface of traits in some fashion.
- Framework Developers is once again hard to quantify, but in general things use more advanced rust concepts like trait objects and dynamic dispatch.


And here is what each person were to gain/lose when having fixed instantiation of components:

- End Users do not see any direct change that is quantifiable
- TE Users only have a fixed configuration to fill out, creating a new deployment requires compilation of a new binary. Losses are needing the infrastructure/knowledge to create the new binaries. Updating the behaviour of a deployment gets harder as more than just the configuration needs to change.
- Framework Users do not see any direct change as writing a component will still be done through the interface of traits in some fashion.
- Framework Developers is hard to quantify due to the question of how abstract the innards are designed.



So overall, I believe that being optional on components is a win for TE users as they get flexibility at low cost. This does make it more complex for framework developers (us), but this is all a finite amount of effort as the Rust concepts are learnable (and teachable).

Reply:
@TheNeikos this is a fair and helpful comparison.

Reply:
I changed my mind on this point. I agree that it will convenient for the agent developers to be able to configure at runtime which components are instantiated and launched.